File: frontend/shared/jupyterlite_mock_widget_example.py
import json
from dataclasses import dataclass, field
from typing import Any

import requests


@dataclass
class MockExamClient:
    base_url: str
    token: str
    assessment_id: int
    attempt_id: int | None = None
    paper: dict[str, Any] | None = None
    responses: dict[int, dict[str, Any]] = field(default_factory=dict)

    @property
    def headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.token}",
            "Content-Type": "application/json",
        }

    def start(self) -> dict[str, Any]:
        r = requests.post(
            f"{self.base_url}/api/mock-exams/{self.assessment_id}/start",
            headers=self.headers,
            timeout=30,
        )
        r.raise_for_status()
        data = r.json()
        self.attempt_id = data["attempt_id"]
        return data

    def fetch_paper(self) -> dict[str, Any]:
        r = requests.get(
            f"{self.base_url}/api/mock-exams/{self.assessment_id}/paper",
            headers=self.headers,
            timeout=30,
        )
        r.raise_for_status()
        self.paper = r.json()
        return self.paper

    def set_response(self, question_id: int, response: dict[str, Any]) -> None:
        self.responses[question_id] = response

    def autosave(self) -> dict[str, Any]:
        if self.attempt_id is None:
            raise RuntimeError("Attempt has not been started")
        payload = {
            "responses": [
                {"question_id": qid, "response": resp}
                for qid, resp in self.responses.items()
            ]
        }
        r = requests.post(
            f"{self.base_url}/api/mock-exams/attempts/{self.attempt_id}/autosave",
            headers=self.headers,
            data=json.dumps(payload),
            timeout=30,
        )
        r.raise_for_status()
        return r.json()

    def submit(self, submitted_payload: dict[str, Any] | None = None) -> dict[str, Any]:
        if self.attempt_id is None:
            raise RuntimeError("Attempt has not been started")
        payload = {
            "submitted_payload": submitted_payload or {
                "submitted_from": "jupyterlite",
                "done": True,
            }
        }
        r = requests.post(
            f"{self.base_url}/api/mock-exams/attempts/{self.attempt_id}/submit",
            headers=self.headers,
            data=json.dumps(payload),
            timeout=30,
        )
        r.raise_for_status()
        return r.json()

File: frontend_old/shared/jupyterlite_mock_widget_example.js
export class JupyterLiteMockExamClient {
  constructor({ baseUrl, token, assessmentId }) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.token = token;
    this.assessmentId = assessmentId;
    this.attemptId = null;
    this.paper = null;
    this.responses = {};
  }

  headers() {
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.token}`,
    };
  }

  async start() {
    const res = await fetch(`${this.baseUrl}/api/mock-exams/${this.assessmentId}/start`, {
      method: "POST",
      headers: this.headers(),
    });
    if (!res.ok) throw new Error(`Failed to start mock exam: ${res.status}`);
    const data = await res.json();
    this.attemptId = data.attempt_id;
    return data;
  }

  async fetchPaper() {
    const res = await fetch(`${this.baseUrl}/api/mock-exams/${this.assessmentId}/paper`, {
      method: "GET",
      headers: this.headers(),
    });
    if (!res.ok) throw new Error(`Failed to fetch paper: ${res.status}`);
    const data = await res.json();
    this.paper = data;
    return data;
  }

  setResponse(questionId, response) {
    this.responses[questionId] = response;
  }

  async autosave() {
    if (!this.attemptId) throw new Error("Attempt not started");
    const payload = {
      responses: Object.entries(this.responses).map(([question_id, response]) => ({
        question_id: Number(question_id),
        response,
      })),
    };
    const res = await fetch(`${this.baseUrl}/api/mock-exams/attempts/${this.attemptId}/autosave`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error(`Autosave failed: ${res.status}`);
    return res.json();
  }

  async submit() {
    if (!this.attemptId) throw new Error("Attempt not started");
    const res = await fetch(`${this.baseUrl}/api/mock-exams/attempts/${this.attemptId}/submit`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        submitted_payload: {
          submitted_from: "jupyterlite",
          done: true,
        },
      }),
    });
    if (!res.ok) throw new Error(`Submit failed: ${res.status}`);
    return res.json();
  }
}
